lib-injection/requirements.json

```diff
@@ -0,0 +1,19 @@
+{
+  "$schema": "https://raw.githubusercontent.com/DataDog/auto_inject/refs/heads/main/preload_go/cmd/library_requirements_tester/testdata/requirements_schema.json",
+  "version": 1,
+  "native_deps": {
+    "glibc": [
+      {
+        "arch": "x86",
+        "supported": true,
+        "min": "2.28"
```

Comment:
2.28 is quite high for x86, do you know what symbol/dependency requires it?

running `nm -D library_file | grep U` will give you that answer

Reply:
The pipeline is using `registry.ddbuild.io/images/mirror/buildpack-deps:buster` as a base: https://github.com/DataDog/dd-trace-rb/blob/1ec390006b42b54a930776096f846c4f20651031/.gitlab/Dockerfile-3.2.2#L4

With which we get a glibc 2.28:

```
$ ldd --version
ldd (Debian GLIBC 2.28-10+deb10u2) 2.28
```

So [gems with native extensions are built against that](https://github.com/DataDog/dd-trace-rb/blob/1ec390006b42b54a930776096f846c4f20651031/.gitlab-ci.yml#L91).

The version requirement comes from there.

This results in the following max version referenced (== min version required):

```
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/datadog-2.4.0/datadog_profiling_loader.3.2.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/datadog-2.4.0/libdatadog_api.3.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/ffi-1.16.3/ffi_c.so: GLIBC_2.27
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/msgpack-1.7.3/msgpack/msgpack.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/datadog-2.4.0/lib/datadog_profiling_loader.3.2.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/datadog-2.4.0/lib/libdatadog_api.3.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/ffi-1.16.3/lib/ffi_c.so: GLIBC_2.27
tmp/arm64/3.2.0/gems/libdatadog-13.1.0.1.0-aarch64-linux/vendor/libdatadog-13.1.0/aarch64-linux-musl/libdatadog-aarch64-alpine-linux-musl/lib/libdatadog_profiling.so: GLIBC_2.0
tmp/arm64/3.2.0/gems/libdatadog-13.1.0.1.0-aarch64-linux/vendor/libdatadog-13.1.0/aarch64-linux/libdatadog-aarch64-unknown-linux-gnu/lib/libdatadog_profiling.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/libddwaf-1.14.0.0.0-aarch64-linux/vendor/libddwaf/libddwaf-1.14.0-linux-aarch64/lib/libddwaf.so: 
tmp/arm64/3.2.0/gems/msgpack-1.7.3/lib/msgpack/msgpack.so: GLIBC_2.17
```

`ffi` is the one using `GLIBC_2.27`. So we can lower it to 2.27 but until we move to building against a lower glibc (which comes with its own set of issues) we're stuck with that. FYI using [manylinux](https://github.com/pypa/manylinux) is under consideration.

Reply:
if it helps you can pick agent build images or the ones used by the injector:
- [centos7 for amd](https://github.com/DataDog/auto_inject/blob/main/.gitlab/Dockerfile.rpm-x64)
- [for arm](https://github.com/DataDog/auto_inject/blob/main/.gitlab/Dockerfile.rpm-arm64)

Given that you download the source codes, won't building ffi on those machines improve your coverage? Or are there specific recent symbols that are required

Reply:
I think [`memfd_create`](https://www.man7.org/linux/man-pages/man2/memfd_create.2.html) needs glibc 2.27. 

It looks like `libffi` has fallback mechanisms if it is absent (see [Memory Usage](https://github.com/libffi/libffi/blob/0859f8431242d5adff21420b9cab538d2af527b5/doc/libffi.texi)), but it is an important security feature.